### PR TITLE
Add dhall-openapi

### PR DESCRIPTION
**Pull request by TristanCacqueray**
Here is a bulk import of dhall-kubernetes-generator plus a few changes to make it build with latest master of dhall.

The import was generated using these commands:

  # Import dhall-kubernetes in dhall-haskell git
  git remote add kubernetes https://github.com/dhall-lang/dhall-kubernetes
  git fetch kubernetes

  # Filter dhall-kubernetes to remove irrelevant changes
  git filter-branch -f --subdirectory-filter dhall-kubernetes-generator kubernetes/master

  # Rename to dhall-openapi
  git checkout -b dhall-kubernetes remotes/kubernetes/master
  mkdir dhall-openapi
  git mv stack.yaml.lock shell.nix src Setup.hs stack.yaml dhall-kubernetes-generator.cabal dhall-openapi/
  git commit -m "Rename dhall-kubernetes-generator into dhall-openapi"

  # Import the branch in dhall-haskell
  git checkout -b dhall-openapi origin/master
  git merge --no-ff --allow-unrelated-histories dhall-kubernetes

### CODE REVIEW

**Comment by TristanCacqueray** [comment]
I am not entirely sure if this is the right place to host the project, but here is a first draft I used to convert yet another openapi file:  https://github.com/TristanCacqueray/dhall-podman

**Comment by TristanCacqueray** [comment]
That is the correct, I merely moved the code out of the dhall-kubernetes project so that it can be build here. More work is needed to decouple the code base from the Kubernertes use-case.

**Comment by sjakobi** [comment]
@TristanCacqueray Is this PR related to https://discourse.dhall-lang.org/t/expense-proposal-turn-dhall-kubernetes-generator-into-a-standalone-openapi-to-dhall-project/278?

What's the motivation for moving it into this git repo instead of keeping it in a separate repo?

**Comment by TristanCacqueray** [comment]
@sjakobi this is related indeed, though I didn't meant to claim the bounty, I rather just wanted to get the project started.

I thought it would be easier to move the code in `dhall-haskell` directly to work on it with master and benefit from the release tooling.

**Comment by Gabriel439** [comment]
@sjakobi: I think it's a good idea to move it into this repository, for two reasons:

* So that it can share the same build/release/CI tooling
* `dhall-kubernetes` is no longer an appropriate home for a more general OpenAPI tool

This does mean that there will be some indirection as we will now need to make changes to this repository and then reference those changes in `dhall-kubernetes`, but I'm fine with doing that

**Comment by Gabriel439** [comment]
@german1608: Yeah, this will need to be fixed to remove and/or generalize the Kubernetes-specific tweaks

**Comment by sjakobi** [comment]
I'm fine with adding `dhall-openapi` to this repo then. I had initially believed that this would be a tool for a rather narrow usecase, similar to `dhall-kubernetes`. Now that I've read a bit more about OpenAPI, it seems to me that it could be more widely applicable.

I'm curious though: What projects (apart from `dhall-kubernetes`?) might want to use `dhall-openapi`?

**Comment by TristanCacqueray** [comment]
@sjakobi projects that uses and publishes OpenAPI (originally known as swagger) could use `dhall-openapi` to generate dhall types so that client request payload can be defined using dhall. I also find it easier to read dhall definition compared to openapi format.

If I understand correctly, the existing code only process `definitions` objects, and perhaps `path` objects could also be converted into dhall for further integration?

**Comment by ari-becker** [comment]
> I'm curious though: What projects (apart from `dhall-kubernetes`?) might want to use `dhall-openapi`?

Well, for starters, we might be able to get [`dhall-prometheus-operator`](https://github.com/coralogix/dhall-prometheus-operator) to be machine-generated instead of being hand-edited. Might not be the greatest example though because it's still coming out of the Kubernetes ecosystem.

In general, I do think that there's a lot of value in something like `dhall-openapi` as the OpenAPI spec includes the schema of request bodies as well as responses, which will help with generating configuration on the basis of API responses. 

A non-OpenAPI example of this pattern: we use [`dhall-terraform-output`](https://github.com/coralogix/dhall-terraform-output) to take AWS identifiers that were generated by Terraform and dynamically insert them into configuration for Kops (i.e. [`dhall-kops`](https://github.com/coralogix/dhall-kops/), specifically [networkID](https://github.com/coralogix/dhall-kops/blob/3caf53826eee29782f6c0b3a8ba8f69c22484fcb/api/v1alpha2/ClusterSpec.dhall#L63) as described in the [documentation](https://kops.sigs.k8s.io/cluster_spec/#networkid)) so that we can program Kops to create a Kubernetes cluster in the VPC which Terraform just programmatically created for us.

So having `dhall-openapi` could take this pattern and make it more generic / widely applicable.

**Comment by Gabriel439** [comment]
@TristanCacqueray: I'd be fine merging this in its present state.  So long as CI passes and it's not in the release script we can continue to iterate on things with subsequent pull requests

File: dhall-openapi/README.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# `dhall-openapi`
+
+## Introduction
+
+This `dhall-openapi` package provides an OpenAPI to Dhall compiler.
```
[the remark this answers is not in the document]
Comment (TristanCacqueray) [reply dhall-openapi/README.md:5]: Next commit adds a note about the development phase. I'm not sure examples can already be provided.

File: dhall-openapi/src/Dhall/Kubernetes/Types.hs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+This applies only to kubernetes resources where ``kind`` and
+``apiVersion`` are statically determined by the resource. See the
+`Kubernetes OpenAPI Spec Readme`:
+https://github.com/kubernetes/kubernetes/blob/master/api/openapi-spec/README.md#x-kubernetes-group-version-kind
+
+For example for a v1 Deployment we have
+
+{ kind = "Deployment"
+, apiVersion = "apps/v1"
+}
```
[the remark this answers is not in the document]
Comment (TristanCacqueray) [reply dhall-openapi/src/Dhall/Kubernetes/Types.hs:85]: It seems like this part will need to be factored out.

**Comment by TristanCacqueray** [comment]
@Gabriel439 alright, I also think this initial work is good to have merged, though I don't mind keeping this open for improvement too. Note that I do not plan on factoring out the kubernetes specifics in the near future.

**Comment by Gabriel439** [comment]
@TristanCacqueray: Yeah, we don't need to factor out the Kubernetes stuff yet

**Pull request merged**